plugins/core/profile.go

```diff
@@ -0,0 +1,341 @@
+	m.Log.Info("adding profile task:", task)
+	if _, exists := m.TraceProfileTasks[task.TaskID]; exists {
+		return
+	}
+	endTime := task.StartTime + int64(task.Duration)*60*1000
+	task.EndTime = endTime
+	task.Status = reporter.Pending
+	m.TraceProfileTasks[task.TaskID] = &task
+	m.TrySetCurrentTask(&task)
+	m.tryStartCPUProfiling()
+}
+
+func (m *ProfileManager) RemoveProfileTask() {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	for k, t := range m.TraceProfileTasks {
+		if t.Status == reporter.Reported || t.EndTime < time.Now().Unix() {
+			delete(m.TraceProfileTasks, k)
+		}
+	}
+}
+
+func (m *ProfileManager) tryStartCPUProfiling() {
+	ok, err := m.profileEvents.ExecuteComplexEvent(CouldProfile)
+	if err != nil {
+		m.Log.Errorf("profile event error:%v", err)
+		return
+	}
+	t := m.TraceProfileTasks[m.currentTask.taskID]
+	if ok && t.Status == reporter.Pending {
+		err := pprof.StartCPUProfile(m.profilingWriter)
+		if err != nil {
+			m.Log.Info("failed to start cpu profiling", err)
+			return
+		}
+		err = m.profileEvents.UpdateBaseEventStatus(IfProfiling, true)
+		if err != nil {
+			m.Log.Errorf("update profile event error:%v", err)
+		}
+		t.Status = reporter.Running
+		go m.monitor()
+	}
+}
+
+func (m *ProfileManager) CheckIfProfileTarget(endpoint string) bool {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	if m.currentTask == nil {
+		return false
+	}
+	return m.currentTask.endpointName == endpoint
+}
+
+func (m *ProfileManager) IfProfiling() bool {
+	return m.profileEvents.BaseEventStatus[IfProfiling]
+}
+
+func (m *ProfileManager) TrySetCurrentTask(task *reporter.TraceProfileTask) {
+	ok, err := m.profileEvents.ExecuteComplexEvent(CouldSetCurTask)
+	if err != nil {
+		m.Log.Errorf("profile event error:%v", err)
+	}
+	if ok {
+		m.generateCurrentTask(task)
+	}
+}
+
+func (m *ProfileManager) generateProfileLabels(traceSegmentID string, minDurationThreshold int64) profileLabels {
+	var l = LabelSet{}
+	l = UpdateTraceLabels(l, SegmentLabel, traceSegmentID, MinDurationLabel, strconv.FormatInt(minDurationThreshold, 10))
+	return profileLabels{
+		labels: &l,
+	}
+}
+
+func (m *ProfileManager) generateCurrentTask(t *reporter.TraceProfileTask) {
+	var c = currentTask{
+		serialNumber:         t.SerialNumber,
+		taskID:               t.TaskID,
+		minDurationThreshold: t.MinDurationThreshold,
+		duration:             t.Duration,
+		endpointName:         t.EndpointName,
+	}
+	m.currentTask = &c
+	err := m.profileEvents.UpdateBaseEventStatus(CurTaskExist, true)
+	if err != nil {
+		m.Log.Errorf("profile event error:%v", err)
+	}
+}
+
+func (m *ProfileManager) TryToAddSegmentLabelSet(traceSegmentID string) {
+	if m.currentTask != nil {
+		c := m.generateProfileLabels(traceSegmentID, m.currentTask.minDurationThreshold)
+		SetGoroutineLabels(c.labels)
+		return
+	}
+}
+
+func (m *ProfileManager) monitor() {
+	done := make(chan struct{})
+	var zeroSince time.Time
+	ticker := time.NewTicker(1 * time.Second)
+	defer ticker.Stop()
+
+	go func() {
+		for range ticker.C {
```

Comment:
If the task has started, and after one second, the endpoint does not trigger any request. The agent will stop the profiling without reaching the end time? This logic may not be correct. 
Also, if the end time is not reached, and you have stopped the profiling, at this time, the agent receives the request that matches the endpoint. Will it restart the profiling? 
So I think for the simplify, you could just stop the profiling when the end time reached.